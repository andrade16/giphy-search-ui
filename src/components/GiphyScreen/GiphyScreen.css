.progress-placeholder {
  height: 4px;
}

.paginate-container {
  display: flex;
  justify-content: center;
}
