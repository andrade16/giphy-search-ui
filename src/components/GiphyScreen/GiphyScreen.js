import React, { useState, useMemo } from "react";
import Search from "../Search/Search";
import CachedSearch from "../Search/CachedSearch";
import { searchGifs } from "../../services/GiphyAPI";
import {ITEMS_PER_PAGE} from "../../utils/constants";
import Container from "@material-ui/core/Container";
import GiphyGrid from "../GiphyGrid/GiphyGrid";
import Paginate from "../Paginate/Paginate";
import EmptyState from "../EmptyState/EmptyState";
import LinearProgress from "@material-ui/core/LinearProgress";
import "./GiphyScreen.css";


const GiphyScreen = () => {
  const [giphyData, setGiphyData] = useState([]);
  const [currentQuery, setCurrentQuery] = useState("");
  const [currentPage, setCurrentPage] = useState(1);
  const [isLoading, setIsLoading] = useState(false);

  const cachedSearch = useMemo(
    () => new CachedSearch(searchGifs, setGiphyData),
    []
  );

  const handleQuerySubmit = (query) => {
    setIsLoading(true);
    cachedSearch.changeQuery(query);
    setCurrentQuery(query);
    setCurrentPage(1);
    if(cachedSearch.cache[query]) {
      setIsLoading(false);
    }
  };

  const pageSelected = (pageNumber) => {
    setIsLoading(true);
    setCurrentPage(pageNumber);
    // calculate offset to fetch next set of data;
    let offset = pageNumber * ITEMS_PER_PAGE - ITEMS_PER_PAGE;
    searchGifs(currentQuery, offset).then((data) => setGiphyData(data));
  };

  const handleLoading = (loading) => {
    setIsLoading(loading);
  };

  const data = giphyData.data || [];

  return (
    <div>
      <Search handleQuery={handleQuerySubmit}/>
      {isLoading && data.length ? (
        <LinearProgress color="secondary" variant="indeterminate" />
      ) : (
        <div className="progress-placeholder"/>
      )}
      <Container maxWidth="md">
        {data && data.length ? (
          <GiphyGrid gifs={data} isLoading={handleLoading} />
        ) : (
          <EmptyState />
        )}
      </Container>
      <div className="paginate-container">
        <Paginate
          currentPage={currentPage}
          itemsPerPage={ITEMS_PER_PAGE}
          totalItems={
            giphyData.pagination ? giphyData.pagination.total_count : []
          }
          pageSelected={pageSelected}
        />
      </div>
    </div>
  );
};

export default GiphyScreen;
