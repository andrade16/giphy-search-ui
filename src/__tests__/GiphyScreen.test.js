import React from 'react';
import renderer from 'react-test-renderer';
import GiphyScreen from '../components/GiphyScreen/GiphyScreen';

describe('GiphyScreen', () => {
    it('renders GiphyScreen correctly', () => {
        const tree = renderer.create(<GiphyScreen/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})
