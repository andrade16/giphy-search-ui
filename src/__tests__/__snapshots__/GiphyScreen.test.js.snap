// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiphyScreen renders GiphyScreen correctly 1`] = `
<div>
  <div
    className="app-bar-container"
  >
    <header
      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary app-bar MuiPaper-elevation4"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <a
          aria-disabled={false}
          aria-label="gif link"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
          href="https://giphy.com/"
          id="gif-button"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          target="_blank"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.5 9H13v6h-1.5zM9 9H6c-.6 0-1 .5-1 1v4c0 .5.4 1 1 1h3c.6 0 1-.5 1-1v-2H8.5v1.5h-2v-3H10V10c0-.5-.4-1-1-1zM19 10.5V9h-4.5v6H16v-2h2v-1.5h-2v-1z"
              />
            </svg>
          </span>
        </a>
        <h6
          className="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
          id="gif-title"
        >
          Giphy Search
        </h6>
        <div
          className="search-container"
        >
          <div
            className="MuiInputBase-root"
            onClick={[Function]}
          >
            <input
              aria-label="search"
              autoComplete="off"
              autoFocus={true}
              className="MuiInputBase-input"
              id="search-input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyUp={[Function]}
              placeholder="Enter a searchâ€¦"
              type="text"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
            disabled={false}
            id="search-button"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Search
              <span
                className="MuiButton-endIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
    </header>
  </div>
  <div
    className="progress-placeholder"
  />
  <div
    className="MuiContainer-root MuiContainer-maxWidthMd"
  >
    <div
      className="empty-state-container"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthMd"
      >
        <div
          className="MuiPaper-root MuiCard-root empty-state-card MuiPaper-outlined MuiPaper-rounded"
        >
          <div
            className="MuiCardContent-root empty-state-card-content"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
              id="empty-state-title"
            >
              No Gifs to Display!
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                id="search-icon"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 13v7H4V6h5.02c.05-.71.22-1.38.48-2H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-5l-2-2zm-1.5 5h-11l2.75-3.53 1.96 2.36 2.75-3.54zm2.8-9.11c.44-.7.7-1.51.7-2.39C20 4.01 17.99 2 15.5 2S11 4.01 11 6.5s2.01 4.5 4.49 4.5c.88 0 1.7-.26 2.39-.7L21 13.42 22.42 12 19.3 8.89zM15.5 9C14.12 9 13 7.88 13 6.5S14.12 4 15.5 4 18 5.12 18 6.5 16.88 9 15.5 9z"
                />
              </svg>
            </p>
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
              id="empty-state-subtext"
            >
              Go ahead and enter a new search to view gifs!
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="paginate-container"
  >
    <div
      className="paginate"
    >
      <nav
        aria-label="pagination navigation"
        className="MuiPagination-root"
      >
        <ul
          className="MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to first page"
              className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded MuiPaginationItem-outlinedPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-label="Go to previous page"
              className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded MuiPaginationItem-outlinedPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded MuiPaginationItem-outlinedPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-label="Go to last page"
              className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded MuiPaginationItem-outlinedPrimary Mui-disabled Mui-disabled"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={-1}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
